components/elm/src/biogeochem/FireDataBaseType.F90

```diff
@@ -0,0 +1,453 @@
+         procedure, public :: FireInit => BaseFireInit                           ! Initialization of Fire
+         procedure, public :: BaseFireInit                                       ! Initialization of Fire
+         procedure, public :: FireInterp                                         ! Interpolate fire data
+         procedure(need_lightning_and_popdens_interface), public, deferred :: &
+              need_lightning_and_popdens ! Returns true if need lightning & popdens
+
+         ! !PRIVATE MEMBER FUNCTIONS:
+         procedure, private :: hdm_init     ! position datasets for dynamic human population density
+         procedure, private :: hdm_interp   ! interpolates between two years of human pop. density file data
+         procedure, private :: lnfm_init    ! position datasets for Lightning
+         procedure, private :: lnfm_interp  ! interpolates between two years of Lightning file data
+         
+     end type fire_base_type
+     !-----------------------------------------------------------------------
+   
+     abstract interface
+        !-----------------------------------------------------------------------
+        function need_lightning_and_popdens_interface(this) result(need_lightning_and_popdens)
+          !
+          ! !DESCRIPTION:
+          ! Returns true if need lightning and popdens, false otherwise
+          !
+          ! USES
+          import :: fire_base_type
+          !
+          ! !ARGUMENTS:
+          class(fire_base_type), intent(in) :: this
+          logical :: need_lightning_and_popdens  ! function result
+          !-----------------------------------------------------------------------
+        end function need_lightning_and_popdens_interface
+     end interface
+   
+     character(len=*), parameter, private :: sourcefile = &
+          __FILE__
+   
+   contains
+   
+     !-----------------------------------------------------------------------
+   !   subroutine BaseFireInit( this, bounds, cnstate_vars )
+     subroutine BaseFireInit( this, bounds, NLFilename )
+       !
+       ! !DESCRIPTION:
+       ! Initialize CN Fire module
+       ! !USES:
+       use shr_infnan_mod  , only : nan => shr_infnan_nan, assignment(=)
+      !  use CNStateType     , only : cnstate_type
+       !
+       ! !ARGUMENTS:
+       class(fire_base_type) :: this
+       type(bounds_type), intent(in)  :: bounds
+       character(len=*), intent(in) :: NLFilename ! Namelist filename
+      !  type(cnstate_type), intent(in) :: cnstate_vars
+
+       !-----------------------------------------------------------------------
+   
+       if ( this%need_lightning_and_popdens() ) then
+          ! Allocate lightning forcing data
+          allocate( this%forc_lnfm(bounds%begg:bounds%endg) )
+          this%forc_lnfm(bounds%begg:) = nan
+          ! Allocate pop dens forcing data
+          allocate( this%forc_hdm(bounds%begg:bounds%endg) )
+          this%forc_hdm(bounds%begg:) = nan
```

Comment:
Rather than initializing to `nan`, could you go ahead and change it to `spval`? I think that will be the recommended practice going forward

Reply:
Will do.  Thanks.

Reply:
@rgknox thinking more about this, `spval` is the positive value equivalent to `fates_unset_r8`.  What are the cases in which we want to use `nan` in fates, versus an unset value?  My assumption is that we want to use this over `nan` generally to make sure that in debug mode, they will end the run?   Do we have an discussion or issue open discussing this on the fates-side? 

@peterdschwartz is there documentation discussing `nan` versus `spval` for e3sm on confluence or in an issue?

Reply:
@glemieux I don't know of details in the documentation regarding this in particular, but as far as I can tell, the approach to `nan` vs `spval` has been to initialize to `nan` at allocation and then if the variable is actually used for computation by namelist, set to `spval` while adding it to the history list. This was done inconsistently and resulted in ELM not being able to be run in debug mode (issue #4820).  The current thinking is that `nan` can be used by developers as a potential debugging tool but shouldn't be hardcoded by default as it just introduces more room for error and complications with marginal benefit.

Reply:
Thanks for that explanation @peterdschwartz